bottlecap/src/appsec/processor/mod.rs

```diff
@@ -0,0 +1,381 @@
+            // We're at capacity, remove the oldest context before proceeding...
+            self.context_buffer.pop_front();
+        }
+
+        // Insert the new context at the back of the buffer.
+        self.context_buffer.push_back(Context::new(
+            rid.to_string(),
+            &mut self.handle,
+            self.waf_timeout,
+        ));
+        // Retrieve a mutable reference to it from the buffer.
+        self.context_buffer
+            .back_mut()
+            .expect("should have at least one element")
+    }
+
+    /// Retrieves the [`Context`] associated with the given request ID, if there
+    /// is one.
+    fn get_context(&self, rid: &str) -> Option<&Context> {
+        self.context_buffer.iter().find(|c| c.rid == rid)
+    }
+
+    /// Retrieves the [`Context`] associated with the given request ID, if there
+    /// is one.
+    fn get_context_mut(&mut self, rid: &str) -> Option<&mut Context> {
+        self.context_buffer.iter_mut().find(|c| c.rid == rid)
+    }
+}
+impl std::fmt::Debug for Processor {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Processor")
+            .field("waf_timeout", &self.waf_timeout)
+            .finish_non_exhaustive()
+    }
+}
+
+/// Error conditions that can arise from calling [`Processor::new`].
+#[derive(Debug)]
+pub enum Error {
+    /// The App & API Protection feature is not enabled
+    FeatureDisabled,
+    /// The WAF builder could not be created (unlikely)
+    WafBuilderCreationFailed,
+    /// The user-configured App & API Protection ruleset file could not be read
+    AppsecRulesError(String, std::io::Error),
+    /// The App & API Protection ruleset could not be parsed from JSON
+    WafRulesetParseError(serde_json::Error),
+    /// The App & API Protection ruleset could not be loaded into the WAF
+    WafRulesetLoadingError(WafOwned<WafMap>),
+    /// The WAF initialization produced a [`None`] handle (this happens when the
+    /// configured ruleset contains no active rule nor processor)
+    WafInitializationFailed,
+}
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::FeatureDisabled => write!(f, "aap: feature is not enabled"),
+            Self::WafBuilderCreationFailed => write!(f, "aap: failed to create WAF builder"),
+            Self::AppsecRulesError(path, e) => {
+                write!(f, "aap: failed to open rules file {path:#?}: {e}")
+            }
+            Self::WafRulesetParseError(e) => write!(f, "aap: failed to parse ruleset: {e}"),
+            Self::WafRulesetLoadingError(diag) => {
+                write!(f, "aap: failed to load ruleset: {diag:?}")
+            }
+            Self::WafInitializationFailed => write!(
+                f,
+                "aap: failed to initialize WAF (is there any active rule or processor in the ruleset?)"
+            ),
+        }
+    }
+}
```

Comment:
Can this be simplified using `thiserror` crate?

Reply:
Ohhhh I did not know about this crate, shineyyy!